Do CA checks to verify authentication by default

brozs opened the pull request:
This is a follow-up to b665b287ff955bbbd9539252ff9f9e2754c3fb48.

An attacker that is able to login into a token could bypass authentication by using its own certificate with any valid signature.

This change makes the default "ca, signature" with the only way to disable CA check by using "no_ca".

This, however, also makes the "none" option disabling CRL and OCSP checks only.

Resolves #80

Jakuje commented on src/common/cert_vfy.c:
```diff
@@ -467,7 +467,7 @@ int verify_certificate(X509 * x509, cert_policy *policy)
 
   /* if neither ca nor crl check are requested skip */
-  if ( (policy->ca_policy==0) && (policy->crl_policy==CRLP_NONE) ) {
+  if ( (!policy->no_ca_policy==0) && (policy->crl_policy==CRLP_NONE) ) {
```
I think it would be more readable to write this as
```suggestion
  if ( (policy->no_ca_policy==1) && (policy->crl_policy==CRLP_NONE) ) {
```

Jakuje commented on src/common/cert_vfy.c:
```diff
@@ -408,7 +408,7 @@ static X509_STORE * setup_store(cert_policy *policy) {
   }
   /* add needed hash dir pathname entries */
-  if ( (policy->ca_policy) && (is_dir(policy->ca_dir)>0) ) {
+  if ( (!policy->no_ca_policy) && (is_dir(policy->ca_dir)>0) ) {
```
I think this would be better readable as
```suggestion
  if ( (policy->no_ca_policy==0) && (is_dir(policy->ca_dir)>0) ) {
```
(similarly the following change)

Jakuje commented on etc/pam_pkcs11.conf.example.in:
```diff
@@ -92,21 +92,22 @@ pam_pkcs11 {
+    #                CRLs
+    # "signature"    Does a signature check to ensure that private
+    #                and public key matches
+    # "no_signature" The only value that disables signature check
```
This needs to be updated also in the documentation. I initially missed that, but I think the xml source is here:

https://github.com/OpenSC/pam_pkcs11/blob/master/doc/pam_pkcs11.xml#L616

If I see right, the `no_signature` part is not documented here either.

Having example files as the only documentation is not very good practice.

brozs committed:
Do CA checks to verify authentication by default

This is a follow-up to b665b287ff955bbbd9539252ff9f9e2754c3fb48.

An attacker that is able to login into a token could bypass
authentication by using its own certificate with any valid signature.

This change makes the default "ca, signature" with the only way to
disable CA check by using "no_ca".

This, however, also makes the "none" option disabling CRL and OCSP
checks only.

Resolves #80

brozs commented:
Updated the commit with the suggestions, and made an attempt to update the documentation, too.

Jakuje commented on src/common/cert_vfy.c:
```diff
@@ -408,7 +408,7 @@ static X509_STORE * setup_store(cert_policy *policy) {
   }
   /* add needed hash dir pathname entries */
-  if ( (policy->ca_policy) && (is_dir(policy->ca_dir)>0) ) {
+  if ( (!policy->no_ca_policy==0) && (is_dir(policy->ca_dir)>0) ) {
```
I think this now changes the meaning and should be as initially suggested in the previous comment:
```suggestion
  if ( (policy->no_ca_policy==0) && (is_dir(policy->ca_dir)>0) ) {
```

brozs committed:
Update src/common/cert_vfy.c

Co-authored-by: Jakub Jelen <jakuje@gmail.com>

Jakuje reviewed (approved):
not sure why the ci fails though

brozs committed:
Fixed extra '>' in doc/pam_pkcs11.xml

wolneykien commented:
@frankmorgner , what do you think about the proposed change? You were the starter! :-)

IMHO, the change seems to be logical. However, if applied, the `"none"` policy becomes the synonym for all _essential_ checks (and the default policy?) and then its name — "none" — sounds quite odd and misleading.

frankmorgner commented:
Thanks for the notification (I am not actively monitoring repository).

Codewise I think the changes are good. Conceptually speaking, I am not sure - I'll discuss that in https://github.com/OpenSC/pam_pkcs11/issues/80

frankmorgner reviewed (approved):
Looks good to me.

I personally would change the explanation of the `none` setting (in manual and example configuration) to something like:

```
# "none"         included for legacy reasons, is equivalent to "ca,signature"
```
This legacy option should then be moved to the bottom of the lists

frankmorgner commented:
I just realized that the code already has support for `configuration.policy.ocsp_policy=OCSP_NONE;`, which cannot be triggered on an individual basis. The same holds for `configuration.policy.crl_policy=CRLP_NONE;`. I would suggest to introduce the new configuration values `no_clr` and `no_ocsp` and then state that `none` translates to `no_crl,no_ocsp`, which hopefully makes it more clear.

wolneykien commented:
> I personally would change the explanation of the `none` setting (in manual and example configuration) to something like:
> 
> ```
> # "none"         included for legacy reasons, is equivalent to "ca,signature"
> ```
> 
> This legacy option should then be moved to the bottom of the lists

Wow, I like the idea.

brozs committed:
Mark cert_policy="none" as legacy in documentation

brozs committed:
Add "no_crl" and "no_ocsp" as cert_policy options

This change allows to individualy set
configuration.policy.ocsp_policy=OCSP_NONE and
configuration.policy.crl_policy=CRLP_NONE by configuration.

Also further clarify the cert_policy="none" option in documentation.

brozs commented:
I have added the `no_crl` and `no_ocsp` options and, hopefully, further improved the clarity of the documentation based on your feedback.

Please let me know if you want to add or change anything else.
